imports/wasi_snapshot_preview1/fs.go

```diff
@@ -643,11 +643,245 @@ func fdRead_shouldContinueRead(n, l uint32, err error) (bool, Errno) {
+func lastDirEntries(dir *internalsys.ReadDir, cookie uint64) (entries []fs.DirEntry, errno Errno) {
+	entryCount := len(dir.Entries)
+	if entryCount == 0 { // there was no prior call
+		if cookie != 0 {
+			errno = ErrnoInval // invalid as we haven't sent that cookie
+		}
+		return
+	}
+
+	countRead := int(dir.CountRead)
+	if int(cookie)-countRead == 1 {
+		// cookie is asking for the next page.
+		dir.Entries = nil
+		return
+	} else if int(cookie)-countRead > 1 {
+		errno = ErrnoInval // invalid as we read that far, yet.
+		return
+	}
+
+	// Get the first absolute position in our window of results
+	firstPos := countRead - entryCount
+	if p := int(cookie) - firstPos; p < 0 {
+		// ReadDir can't read backwards, so we can only implement positions
+		// we still have data for.
+		errno = ErrnoNosys // unimplemented
+		return
+	} else if p == entryCount {
+		return // at the position of the next page.
+	} else if p > 0 {
+		entries = dir.Entries[p:] // truncate so to avoid large lists.
+	} else if dir.Entries != nil {
+		entries = dir.Entries
+	}
+	return
+}
+
+// direntSize is the size of the dirent struct, which should be followed by the
+// length of a file name.
+const direntSize = uint32(24)
+
+// maxDirents returns the maximum count and total entries that can fit in
+// maxLen bytes.
+//
+// truncatedEntryLen is the amount of bytes past bufLen needed to write the
+// next entry. We have to return bufused == bufLen unless the directory is
+// exhausted.
+//
+// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#fd_readdir
+// See https://github.com/WebAssembly/wasi-libc/blob/659ff414560721b1660a19685110e484a081c3d4/libc-bottom-half/cloudlibc/src/libc/dirent/readdir.c#L44
+func maxDirents(entries []fs.DirEntry, bufLen uint32) (bufused, direntCount, truncatedEntryLen uint32) {
+	lenRemaining := bufLen
+	for _, e := range entries {
+		nameLen := uint32(len(e.Name()))
+		entryLen := direntSize + nameLen
+
+		if lenRemaining == 0 {
+			// We've hit exactly the end of bufLen, so cannot write the next
+			// entry. A caller who wants more will retry.
+			break
+		}
+
+		if lenRemaining < entryLen {
+			// We haven't room to write the entry, and docs say to write the
+			// header. This helps especially when there is an entry with a very
+			// long filename. Ex if bufLen is 4096 and the filename is 4096,
+			// we need to write direntSize(24) + 4096 bytes to write the entry.
+			// In this case, we only write up to direntSize(24) to allow the
+			// caller to resize.
+
+			// bufused != bufLen means more entries exist, which is the case
+			// when the next entry is larger than bytes remaining.
+			bufused = bufLen
+
+			// When the next entry is truncated, only write the dirent.
+			if lenRemaining < direntSize {
+				truncatedEntryLen = direntSize - lenRemaining
+			}
+			break
+		}
+
+		// At this point, this entry can fit within bufLen.
+		bufused += entryLen
+		lenRemaining -= entryLen
```

Comment:
could you teach me why this won't be negative..

Reply:
we did `if lenRemaining < entryLen` before this line so worst case this is zero right?

Reply:
it can be negative if the name size is a very large number. I think we can skip in that case, though testing it would be extremely expensive I think

Reply:
I think I have a way to test it.. we are only reading the size of each entry and nothing else

Reply:
got a simper way, nm